src/Workflow/MoveJatsAssets.php

```diff
@@ -0,0 +1,139 @@
+use GuzzleHttp\Psr7\UriResolver;
+use InvalidArgumentException;
+use League\Flysystem\AdapterInterface;
+use League\Flysystem\FilesystemInterface;
+use Libero\ContentApiBundle\Model\PutTask;
+use Psr\Http\Message\ResponseInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\File\MimeType\ExtensionGuesser;
+use Symfony\Component\Workflow\Event\Event;
+use UnexpectedValueException;
+use function count;
+use function explode;
+use function GuzzleHttp\Promise\all;
+use function GuzzleHttp\Psr7\uri_for;
+use function implode;
+use function in_array;
+use function is_callable;
+use function is_resource;
+use function Libero\ContentApiBundle\stream_hash;
+use function preg_match;
+use function sprintf;
+
+final class MoveJatsAssets implements EventSubscriberInterface
+{
+    private const HAS_MIMETYPE_ATTRIBUTE = [
+        'graphic',
+        'inline-graphic',
+        'inline-media',
+        'inline-supplementary-material',
+        'media',
+        'supplementary-material',
+    ];
+
+    private $client;
+    private $filesystem;
+    private $origin;
+
+    public function __construct(string $origin, FilesystemInterface $filesystem, ClientInterface $client)
+    {
+        $this->origin = $origin;
+        $this->filesystem = $filesystem;
+        $this->client = $client;
+
+        if (!is_callable([$filesystem, 'getPublicUrl'])) {
+            throw new InvalidArgumentException('Requires the public URL plugin');
+        }
+    }
+
+    public static function getSubscribedEvents() : array
+    {
+        return [
+            'workflow.libero.content_store.put.transition.manipulate' => 'onManipulate',
+        ];
+    }
+
+    public function onManipulate(Event $event) : void
+    {
+        /** @var PutTask $task */
+        $task = $event->getSubject();
+        $document = $task->getDocument();
+
+        $document->registerNamespace('jats', 'http://jats.nlm.nih.gov');
+        $document->registerNamespace('xlink', 'http://www.w3.org/1999/xlink');
+
+        /** @var array<Element> $assets */
+        $assets = $document('//jats:article//jats:*[@xlink:href]');
+
+        $promises = [];
+
+        foreach ($assets as $asset) {
+            $uri = uri_for($asset->getAttribute('xlink:href') ?? '');
+
+            if ($asset->baseURI && !$uri->getScheme()) {
+                $uri = UriResolver::resolve(uri_for($asset->baseURI), $uri);
+            }
+
+            $uri = UriNormalizer::normalize($uri);
+
+            if (!$uri->getScheme() || 0 === preg_match($this->origin, (string) $uri)) {
+                continue;
+            }
+
+            $promises[] = $this->client
+                ->requestAsync('GET', $uri, ['http_errors' => true])
+                ->then(
+                    function (ResponseInterface $response) use ($asset, $task) {
+                        $contentType = explode(';', $response->getHeaderLine('Content-Type'), 2)[0] ?? '';
```

Comment:
isn't this likely to be garbage data (`Content-Type` from a random webserver used to temporarily host files)?

There are ways to force this on S3, but they make the adapter side even more specific to it. The defaults don't seem encouraging:
```
$ curl -v https://elife-cdn.s3.amazonaws.com/ping_fastly.txt -o /dev/null 2>&1 | grep Content-Type
< Content-Type: application/xml
```

Reply:
Urgh. Can guess it from the content itself, but think that would involve writing it to disk.

Reply:
Maybe a stream wrapper can work...

Reply:
https://elife-cdn.s3.amazonaws.com/ping_fastly.txt is an XML error page...

Reply:
https://elife-cdn.s3.amazonaws.com/ping-fastly has `text/plain`.

Reply:
Seems like S3 sets a default based on the file extension, or `binary/octet-stream` otherwise. Think it's fine to use it. We can add guessing based on the content later.

Reply:
That's reasonable if we stick to consistent extensions :+1:

Reply:
(but tied to S3 still)

Reply:
There's no tie to S3 here, it's just HTTP.

Reply:
Elaborating: to an HTTP origin that sets reasonable `Content-Type` headers; S3 and probably static webservers should cover this, though it should be made explicit in the contract that the content API presents to the outside world. There's no lock in here to AWS specific things like authentication headers.